File: Easy/0067-add-binary/addBinary.py
class Solution:
    def addBinary(self, a: str, b: str) -> str:
        # Pointers starting from the least significant bit (end of each string)
        i = len(a) - 1
        j = len(b) - 1

        carry = 0
        out = []  # we'll store result bits here (in reverse order for efficiency)

        # Continue while there are bits left in a or b, or we still have a carry
        while i >= 0 or j >= 0 or carry:
            total = carry

            # Add current bit from a (if any)
            if i >= 0:
                total += ord(a[i]) - ord('0')
                i -= 1

            # Add current bit from b (if any)
            if j >= 0:
                total += ord(b[j]) - ord('0')
                j -= 1

            # The resulting bit is total % 2, new carry is total // 2
            out.append(str(total & 1))
            carry = total >> 1

        # out is reversed (LSB first), so reverse and join into a string
        out.reverse()
        return "".join(out)

File: Easy/0067-add-binary/addBinary.cpp
#include <string>
#include <algorithm>

class Solution {
public:
    std::string addBinary(std::string a, std::string b) {
        int i = (int)a.size() - 1;  // index from the end of a
        int j = (int)b.size() - 1;  // index from the end of b
        int carry = 0;

        std::string res;
        res.reserve(std::max(a.size(), b.size()) + 1); // avoid repeated reallocations

        // Add bits from right to left (LSB to MSB)
        while (i >= 0 || j >= 0 || carry) {
            int sum = carry;

            // Convert current char bit to int (0/1) and add it
            if (i >= 0) sum += (a[i--] - '0');
            if (j >= 0) sum += (b[j--] - '0');

            // Current result bit is sum % 2, next carry is sum / 2
            res.push_back(char('0' + (sum & 1)));
            carry = (sum >> 1);
        }

        // We built the result reversed, so reverse it to correct order
        std::reverse(res.begin(), res.end());
        return res;
    }
};